### PR TITLE
Try & simplify Registry vs Source confusion

### PR DESCRIPTION
**Pull request by dwijnand**
Refs #3006.

I tried to go further and replace references of `Registry` with `PackageRegistry`, but I ran into lifetime issue in `ops::resolve_with_previous` which I think has to do with no longer passing `PackageRegistry` as a `Registry` trait object (with a new and distinct lifetime).

### CODE REVIEW

**Comment by dwijnand** [comment]
I just had a thought: is it possible to delegate from one trait to the other, rather than have both traits delegate to the struct?

File: src/cargo/core/registry.rs [hunk 1]
```diff
@@ -21,28 +21,12 @@ pub trait Registry {
...
-    /// Returns whether or not this registry will return summaries with
-    /// checksums listed.
-    fn supports_checksums(&self) -> bool;
-
-    /// Returns whether or not this registry will return summaries with
-    /// the `precise` field in the source id listed.
-    fn requires_precise(&self) -> bool;
 }
 
 impl<'a, T: ?Sized + Registry + 'a> Registry for Box<T> {
```
Comment (matklad) [inline src/cargo/core/registry.rs]: We can remove this impl altogether.

File: src/cargo/sources/directory.rs [hunk 2]
```diff
@@ -46,12 +55,13 @@ impl<'cfg> Debug for DirectorySource<'cfg> {
 
 impl<'cfg> Registry for DirectorySource<'cfg> {
```
Comment (matklad) [inline src/cargo/sources/directory.rs]: I think we now can remove all `impl Registry for XSource` blocks.

**Comment by matklad** [comment]
Looks like a good start, but now I believe we can remove a whole bunch of useless impls!

**Comment by dwijnand** [comment]
You're right, @matklad! I hadn't realised.

That's much better. My first net 🔻 PR! 🎉

File: src/cargo/core/registry.rs [hunk 3]
```diff
@@ -703,13 +673,5 @@ pub mod test {
                 Ok(())
```
[the remark this answers is not in the document]
Comment (matklad) [reply src/cargo/core/registry.rs]: Wow, this is dead code indeed! Let's remove it!

**Comment by alexcrichton** [comment]
👍

**Comment by matklad** [comment]
@bors r+

Thanks!

**Pull request merged**